app/code/Magento/Checkout/Model/ShippingInformationManagement.php

```diff
@@ -177,7 +177,7 @@ public function saveAddressInformation(
 
         $shippingAddress = $quote->getShippingAddress();
 
-        if (!$shippingAddress->getShippingRateByCode($shippingAddress->getShippingMethod())) {
+        if (!$shippingAddress->getShippingRateByCode($shippingAddress->getShippingMethod()) && !$quote->getIsVirtual()) {
```

Comment:
I believe it would be more logical to swap condition (so that we do not even look for shipping rates in case of virtual quote).